core/src/main/java/hudson/util/RingBufferLogHandler.java

```diff
@@ -86,13 +75,36 @@ public static int getDefaultRingBufferSize() {
     }
 
     @Override
-    public synchronized void publish(LogRecord record) {
-        int len = records.length;
-        records[(start + size) % len] = new LogRecordRef(record);
-        if (size == len) {
-            start = (start + 1) % len;
-        } else {
-            size++;
+    public void publish(LogRecord record) {
+        if (record == null) {
+            return;
+        }
+        if (JenkinsJVM.isJenkinsJVM() && record.getParameters() != null) {
```

Comment:
Good that we are not reintroducing the deadlocks experienced previously when trying to eagerly format log records on agent JVMs. My understanding is this PR eliminates `LogRecordRef`, which means that someone could now again (as they could before #6018) set up a custom logger that records messages from an agent JVM and whose parameter references may last beyond the end of the build. So in other words, this PR returns us to the pre-#6018 status quo with the addition of eager formatting on the controller side (but without eager formatting on the agent side, which carries additional implementation difficulties). This seems like a net improvement as long as the lack of agent-side support does not interfere with things like BOM/PCT testing (which was the original motivation behind working on this in #6018).

Reply:
> we are not reintroducing the deadlocks experienced previously

That I know of. It is particularly difficult to reason about `Logger` behavior because it can be called from sensitive contexts such as class loading.

> the lack of agent-side support

Exactly, the assumption here is that controller JVMs are likely to last much longer than agent JVMs; a crash of an agent JVM is more easily recovered; and the known scenario motivating the original change (a giant `GroovyClassLoader` in a message format) was on the controller, not the agent.

Reply:
> That I know of. It is particularly difficult to reason about Logger behavior because it can be called from sensitive contexts such as class loading.

I suppose there is a risk of deadlock when loading the `JenkinsJVM` class, just as there previously was with `LogRecordRef`. If that manifests itself in practice, I suppose we could solve that issue by preloading it as we did with `LogRecordRef`.